src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs

```diff
@@ -43548,4 +43551,1989 @@ static void Main()
+    {
+        var x = new { A = 1, B = 2 };
+        var y = Extend(x);
+        y.Method();
+    }
+
+    public static E<T> Extend<T>(T t) => throw null; // => (E<T>)t;
+}
+""";
+
+        var comp = CreateCompilation([src, ExtensionErasureAttributeDefinition], options: TestOptions.DebugExe);
+        // PROTOTYPE verify execution once conversion support is added
+        var verifier = CompileAndVerify(comp/*, expectedOutput: "ran"*/).VerifyDiagnostics();
+        verifier.VerifyIL("D.Main", """
+{
+  // Code size       25 (0x19)
+  .maxstack  2
+  .locals init (<>f__AnonymousType0<int, int> V_0, //x
+                <>f__AnonymousType0<int, int> V_1) //y
+  IL_0000:  nop
+  IL_0001:  ldc.i4.1
+  IL_0002:  ldc.i4.2
+  IL_0003:  newobj     "<>f__AnonymousType0<int, int>..ctor(int, int)"
+  IL_0008:  stloc.0
+  IL_0009:  ldloc.0
+  IL_000a:  call       "E<<anonymous type: int A, int B>> D.Extend<<anonymous type: int A, int B>>(<anonymous type: int A, int B>)"
+  IL_000f:  stloc.1
+  IL_0010:  ldloca.s   V_1
+  IL_0012:  call       "void E<<anonymous type: int A, int B>>.Method(ref <anonymous type: int A, int B>)"
+  IL_0017:  nop
+  IL_0018:  ret
+}
+""");
+    }
+
+    [Fact(Skip = "PROTOTYPE assertion in NullableWalker when accessing anonymous type property from underlying type")]
+    public void TypeReference_MethodBody_AnonymousType_AccessAnonymousTypeMembers()
+    {
+        var src = """
+public explicit extension E<T> for T { }
+
+public class D
+{
+    public static void Main()
+    {
+        var x = new { A = 1, B = 2 };
+        var y = Extend(x);
+        System.Console.Write(y.A);
+    }
+
+    public static E<T> Extend<T>(T t) => (E<T>)t;
+}
+""";
+
+        var comp = CreateCompilation([src, ExtensionErasureAttributeDefinition], options: TestOptions.DebugExe);
+        // PROTOTYPE verify execution once conversion support is added
+        var verifier = CompileAndVerify(comp/*, expectedOutput: "ran"*/).VerifyDiagnostics();
+        verifier.VerifyIL("D.Main", """
+{
+  // Code size       25 (0x19)
+  .maxstack  2
+  .locals init (<>f__AnonymousType0<int, int> V_0, //x
+                <>f__AnonymousType0<int, int> V_1) //y
+  IL_0000:  nop
+  IL_0001:  ldc.i4.1
+  IL_0002:  ldc.i4.2
+  IL_0003:  newobj     "<>f__AnonymousType0<int, int>..ctor(int, int)"
+  IL_0008:  stloc.0
+  IL_0009:  ldloc.0
+  IL_000a:  call       "E<<anonymous type: int A, int B>> D.Extend<<anonymous type: int A, int B>>(<anonymous type: int A, int B>)"
+  IL_000f:  stloc.1
+  IL_0010:  ldloca.s   V_1
+  IL_0012:  call       "void E<<anonymous type: int A, int B>>.Method(ref <anonymous type: int A, int B>)"
+  IL_0017:  nop
+  IL_0018:  ret
+}
+""");
+    }
+
+    [Fact]
+    public void TypeReference_Typeof_Simple()
+    {
+        var src = """
+System.Console.Write(D.M().ToString());
+
+public explicit extension E for U { }
+
+public class U { }
+public class D
+{
+    public static System.Type M()
+    {
+        return typeof(E);
```

Comment:
Should we test `typeof` and `default` that references an extension type in an attribute?

Reply:
Good idea. I'll add tests for `typeof` in attribute here, but `default` will be in a follow-up PR (focused on expressions and conversions).